### PR TITLE
Add Master Throttling Metric

### PR DESCRIPTION
**Pull request by dhwanilpatel**
*Description of changes:*
Add Metrics for Master Throttling of Pending tasks. It will publish two metric 1) Total throttled tasks at master node 2) Number of task on which data node is actively retrying. This feature is yet to be contributed to Opensource ES. I have added the check for availability of this feature in collector so building with or without this feature will pass. If this feature is not available then collector will simply return. 

PR of Performance Analyzer Plugin : https://github.com/opendistro-for-elasticsearch/performance-analyzer/pull/217

*Tests:*

Testing

1. /gradlew test Successful
```
SUCCESS: Executed 455 tests in 40m 43s (1 skipped)

BUILD SUCCESSFUL in 41m 8s
11 actionable tasks: 8 executed, 3 up-to-date
```
2. Tested in test domain by replacing JAR.

Tmp file
```
^master_throttling_metrics
{"current_time":1602617137529}
{"Data_RetryingPendingTasksCount":0,"Master_ThrottledPendingTasksCount":0}$
```

Verified metric from table
```
sqlite> select * from Master_ThrottledPendingTasksCount;
0.0|0.0|0.0|0.0
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by yojs** [comment]
The CI isn't passing. There are some spot bug failures. You can rerun the spotbugs as 
`./gradlew spotbugsMain`

And then point your browser to `<project dir>/build/reports/spotbugs/main.html`

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/metrics/AllMetrics.java [hunk 1]
```diff
@@ -821,6 +821,27 @@ public String toString() {
     }
   }
 
+  public enum MasterThrottlingValue implements MetricValue {
+    MASTER_THROTTLED_PENDING_TASK_COUNT(MasterThrottlingValue.Constants.THROTTLED_PENDING_TASK_COUNT),
```
[the remark this answers is not in the document]
Comment (dhwanilpatel) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/metrics/AllMetrics.java:825]: Added Javadoc.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/reader/MasterThrottlingMetricsEventProcessor.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    public void processEvent(Event event) {
+        String[] lines = event.value.split(System.lineSeparator());
+        for (String line : lines) {
+            Map<String, String> masterThrottlingMap = extractEntryData(line);
+            if (!masterThrottlingMap.containsKey(PerformanceAnalyzerMetrics.METRIC_CURRENT_TIME)) {
+                handle.bind(
+                        Long.parseLong(masterThrottlingMap.get(
+                                AllMetrics.MasterThrottlingValue.DATA_RETRYING_TASK_COUNT.toString())),
+                        Long.parseLong(masterThrottlingMap.get(
+                                AllMetrics.MasterThrottlingValue.MASTER_THROTTLED_PENDING_TASK_COUNT.toString())));
```
[the remark this answers is not in the document]
Comment (dhwanilpatel) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/reader/MasterThrottlingMetricsEventProcessor.java:73]: Added try/catch.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/reader/MasterThrottlingMetricsEventProcessor.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.Map;
+import java.util.NavigableMap;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jooq.BatchBindStep;
+
+public class MasterThrottlingMetricsEventProcessor implements EventProcessor {
+    private static final Logger LOG = LogManager.getLogger(MasterThrottlingMetricsEventProcessor.class);
+    private MasterThrottlingMetricsSnapshot masterThrottlingMetricsSnapshot;
```
[the remark this answers is not in the document]
Comment (dhwanilpatel) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/reader/MasterThrottlingMetricsEventProcessor.java:21]: Corrected.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/reader/MasterThrottlingMetricsEventProcessor.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    public void commitBatchIfRequired() {
+        if (handle.size() > BATCH_LIMIT) {
+            handle.execute();
+            handle = masterThrottlingMetricsSnapshot.startBatchPut();
+        }
+    }
+
+    static Map<String, String> extractEntryData(String line) {
+        try {
+            Map<String, String> map = MAPPER.readValue(line, TYPE_REF);
```
[the remark this answers is not in the document]
Comment (dhwanilpatel) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/reader/MasterThrottlingMetricsEventProcessor.java:93]: Corrected.

**Review by aditjind — DISMISSED** [review]

**Comment by aditjind** [comment]
There are some checkstyle errors in the CI. Fix them and merge the changes in the RCA packages first since PA depends on RCA to build (for Constants etc.).

**Review by yojs — APPROVED** [review]

**Review by aditjind — APPROVED** [review]

**Pull request merged**